redriver/src/main/java/com/flipkart/flux/redriver/service/RedriverService.java

```diff
@@ -0,0 +1,108 @@
+public class RedriverService {
+
+    private static final Logger logger = LoggerFactory.getLogger(RedriverService.class);
+    private static final String scheduledExectorSvcName = "redriver-batch-read-executor-svc";
+
+    private Integer batchReadInterval;
+    private Integer batchSize;
+    private MessageManagerService messageService;
+    private ScheduledFuture scheduledFuture;
+    private final RedriverRegistry redriverRegistry;
+    private final InstrumentedScheduledExecutorService scheduledExecutorService;
+
+    @Inject
+    public RedriverService(MessageManagerService messageService,
+                           RedriverRegistry redriverRegistry,
+                           @Named("redriver.batchread.intervalms") Integer batchReadInterval,
+                           @Named("redriver.batchread.batchsize") Integer batchSize) {
+        this.redriverRegistry = redriverRegistry;
+        this.batchReadInterval = batchReadInterval;
+        this.batchSize = batchSize;
+        this.messageService = messageService;
+
+        ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
+        executor.setRemoveOnCancelPolicy(true);
+        scheduledExecutorService =
+                new InstrumentedScheduledExecutorService(Executors.unconfigurableScheduledExecutorService(executor),
+                        SharedMetricRegistries.getOrCreate(METRIC_REGISTRY_NAME), scheduledExectorSvcName);
+    }
+
+    public void start() {
+        final long initialDelay = 10000l;
+        synchronized (this) {
+            if (scheduledFuture == null || scheduledFuture.isDone()) {
+                scheduledFuture = scheduledExecutorService.scheduleWithFixedDelay(() -> {
+                    try {
+                        redrive();
+                    }
+                    catch(Exception e) {
+                        logger.error("Error while running redrive", e);
+                    }
+                }, initialDelay, batchReadInterval, TimeUnit.MILLISECONDS);
+                logger.info("RedriverService started");
+            }
+        }
+    }
+
+    public void stop() {
+        synchronized (this) {
+            scheduledFuture.cancel(false);
+            logger.info("RedriverService stopped");
+        }
+    }
+
+    public Boolean isRunning() {
+        return scheduledFuture != null && (!scheduledFuture.isDone());
+    }
+
+    private void redrive() {
+        int offset = 0;
+        List<ScheduledMessage> messages;
+        long now = System.currentTimeMillis();
+        do {
+            messages = messageService.retrieveOldest(offset, batchSize);
+            messages.stream().filter(e -> e.getScheduledTime() < now).forEach(e -> redriverRegistry.redriveTask(e.getTaskId()));
+            offset += batchSize;
+            // get next batch if we found batchSize tasks and all were redriven
+        } while (messages.size() == batchSize && messages.get(messages.size() - 1).getScheduledTime() < now);
+    }
```

Comment:
you mean, if there are only few total msgs in db, then we clear them up regardless of their timings. OR if we found few msgs that can be redriven then we should do that. the latter is already happening.